File: Couchbase.Lite/Couchbase.Lite/Exceptions/CouchbaseLiteException.cs
using System;

namespace Couchbase.Lite {

    public class CouchbaseLiteException : ApplicationException {

        internal StatusCode Code { get; set; }

        public CouchbaseLiteException (Exception innerException, StatusCode code) : base(String.Format("Database error: {0}", code), innerException) { Code = code; }

        public CouchbaseLiteException (Exception innerException, Status status) : this(innerException, status.GetCode()) { Code = status.GetCode(); }

        public CouchbaseLiteException (StatusCode code) : base(String.Format("Database error: {0}", code)) { Code = code; }

        public CouchbaseLiteException (String message, StatusCode code) : base(message) { Code = code; }

        public CouchbaseLiteException (String message) : base(message) {  }

        public CouchbaseLiteException (String messageFormat, params Object[] values)
            : base(String.Format(messageFormat, values)) {  }

        public Status GetCBLStatus ()
        {
            return new Status(Code);
        }
    }

}

File: src/main/sharpen/java/com/couchbase/lite/CouchbaseLiteException.java
package com.couchbase.lite;

public class CouchbaseLiteException extends Exception {

    private Status status;

    public CouchbaseLiteException(int statusCode) {
        this.status = new Status(statusCode);
    }

    public CouchbaseLiteException(Status status) {
        this.status = status;
    }

    public CouchbaseLiteException(String detailMessage, Status status) {
        super(detailMessage);
        this.status = status;
    }

    public CouchbaseLiteException(String detailMessage, int statusCode) {
        this(detailMessage, new Status(statusCode));
    }

    public CouchbaseLiteException(String detailMessage, Throwable throwable, Status status) {
        super(detailMessage, throwable);
        this.status = status;
    }

    public CouchbaseLiteException(Throwable throwable, Status status) {
        super(throwable);
        this.status = status;
    }

    public CouchbaseLiteException(Throwable throwable, int statusCode) {
        super(throwable);
        this.status = new Status(statusCode);
    }

    public Status getCBLStatus() {
        return status;
    }

    // TODO: override toString to print out status code

}
